### ISSUE TITLE
Duplicated useId values

### ISSUE THREAD

**Issue by grefrit** [preactjs/preact#3757]
- [ ] Check if updating to the latest Preact version resolves the issue

**Describe the bug**
Sometimes useId() values are duplicated. I saw this strange behavior in @headlessui/react components. Headlessui uses useId only when available (to support React 17). In development I use React 17 and everything looks normal, but in production I'm switching React to Preact 10.11.1 and see this in some situations
<img width="692" alt="image" src="https://user-images.githubusercontent.com/45934058/194133276-186182da-51c8-431f-97c2-1b14335ca07b.png">
As you can see values are duplicated. I can confirm that is it issue in Preact, not headlessui, because it works when headlessui not using useId hook. It is very strange that this does not happen in all elements, but only in some. I said about headlessui, because I'm not using useId hook myself, so I don't know how to reproduce this behavior. I will try to do my best and provide some reproduction later. Also I can add that this strange behavior happens in this two elements
<img width="441" alt="image" src="https://user-images.githubusercontent.com/45934058/194134838-ea85bc9b-f563-482b-b596-2a1b2db5d5cf.png">

So to sum up.

1. This happens sometimes, I can't find some logic when it happens, it just duplicates value, very strange
2. I told about headlessui, because I don't use useId hook myself and I don't know how to reproduce this

**To Reproduce**
~~Try to provide soon~~
UPD: https://github.com/preactjs/preact/issues/3757#issuecomment-1268982257

**Expected behavior**
Values of useId should be unique

**Comment by marvinhagemeister** [comment]
Found a minimal reproduction case:

```jsx
import { useId } from "preact/hooks";

function Foo() {
  const id = useId();
  return <p>{id}</p>;
}

export function App() {
  return (
    <div>
      <Foo />
      <>
        <Foo />
      </>
    </div>
  );
}
```

**Comment by JoviDeCroock** [comment]
Hmm that makes sense, we have a condition to skip fragments. We should disable that and ensure we leverage a root-wrapping fragment in render to string

**Comment by marvinhagemeister** [comment]
@JoviDeCroock currently trying that out, but getting mismatches with render to string. My hunch so far is that the implicit Fragment wrapping we do for array children is different from core.

**Comment by marvinhagemeister** [comment]
Here is a test case where I can force mismatches with RTS:

```jsx
function Foo() {
	const id = useId();
	return <p>{id}</p>;
}

function Bar(props) {
	return props.children;
}

function App() {
	return (
		<Bar>
			<Foo />
			<Fragment>
				<Foo />
			</Fragment>
		</Bar>
	);
}
```

**Comment by JoviDeCroock** [comment]
Concretely to solve this I think we need to:

- [start from the parent](https://github.com/preactjs/preact-render-to-string/blob/master/src/index.js#L64)
- [transform this into fragment wrapping](https://github.com/preactjs/preact-render-to-string/blob/master/src/index.js#L203)
- [remove the fragment bailout](https://github.com/preactjs/preact/blob/master/hooks/src/index.js#L33)

I might be forgetting a case with the implicit fragment wrapping though 😅 don't have a lot of time this week but can look over the weekend.

**Comment by Armster15** [comment]
Unless if I'm doing something wrong, this issue seems to still be occurring even with 10.11.12.
Like grefrit, I too am using Headless UI and the IDs generated by Preact seem to be non unique.

A reproducible repository: https://github.com/Armster15/preact-duplicate-id
Live Demo: https://preact-duplicate-id.vercel.app/
An image showing the duplicate IDs:
<img width="517" alt="Screenshot 2022-11-12 at 2 04 11 PM" src="https://user-images.githubusercontent.com/49625265/201496345-6259f1c3-e9fd-4652-b305-435f2c6deef5.png">

For others coming across this issue, downgrading to Preact 10.10.6 worked for me. This release does not include a `useId` hook, and as such, [Headless UI will provide its own implementation of the hook](https://github.com/tailwindlabs/headlessui/blob/main/packages/%40headlessui-react/src/hooks/use-id.ts#L16), which seems to be working as intended.

**Comment by marvinhagemeister** [comment]
@Armster15 That's because we haven't cut a new release yet. Will do so momentarily.
